src/read.rs

```diff
@@ -425,6 +426,55 @@ impl<'a> SliceRead<'a> {
         }
     }
 
+    #[inline(always)]
+    fn skip_to_escape(&mut self, forbid_control_characters: bool) {
+        let rest = &self.slice[self.index..];
+        let end = self.index + memchr::memchr2(b'"', b'\\', rest).unwrap_or(rest.len());
+
+        if !forbid_control_characters {
+            self.index = end;
+            return;
+        }
+
+        // We now wish to check if the chunk contains a byte in range 0x00..=0x1F. Ideally, this
+        // would be integrated this into the memchr2 check above, but memchr does not support this
+        // at the moment. Therefore, use a variation on Mycroft's algorithm [1] to provide
+        // performance better than a naive loop. It runs faster than just a single memchr call on
+        // benchmarks and is faster than both SSE2 and AVX-based code, and it's cross-platform, so
+        // probably the right fit.
+        // [1]: https://groups.google.com/forum/#!original/comp.lang.c/2HtQXvg7iKc/xOJeipH6KLMJ
+        const STEP: usize = mem::size_of::<usize>();
+
+        // Moving this to a local variable removes a spill in the hot loop.
+        let mut index = self.index;
+
+        if self.slice.len() >= STEP {
+            while index < end.min(self.slice.len() - STEP + 1) {
+                // We can safely overread past end in most cases. This ensures that SWAR code is
+                // used to handle the tail in the hot path.
+                const ONE_BYTES: usize = usize::MAX / 255;
+                let chars = usize::from_ne_bytes(self.slice[index..][..STEP].try_into().unwrap());
+                let mask = chars.wrapping_sub(ONE_BYTES * 0x20) & !chars & (ONE_BYTES << 7);
+
+                if mask != 0 {
+                    index += mask.trailing_zeros() as usize / 8;
```

Comment:
This `trailing_zeros` is only correct if the trailing (least significant) bytes are at lower memory addresses, i.e. little endian architecture. On big endian we would need to use `leading_zeros`, or avoid `from_ne_bytes` above.

You can reproduce the failure using Miri with a big endian target: `cargo miri test --test test --target powerpc64-unknown-linux-gnu test_control_character_search` (comment out the `serde_stacker` dev-dependency if this gives an error about "powerpc-linux-gnu-gcc").

Reply:
Right, I've only ever tested code like this on LE so totally forgot about this. Thanks! I'll see what I can do about it

Reply:
Fixed by using `leading_zeros`. I'm assuming that load + lzcnt is faster than byteswapped load + tzcnt on all platforms.